controllers/runnerpool_controller.go

```diff
@@ -0,0 +1,229 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const runnerPoolFinalizer = "actions.cybozu.com/runnerpool"
+
+// RunnerPoolReconciler reconciles a RunnerPool object
+type RunnerPoolReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	organizationName string
+}
+
+// NewRunnerPoolReconciler creates RunnerPoolReconciler
+func NewRunnerPoolReconciler(
+	client client.Client,
+	log logr.Logger,
+	scheme *runtime.Scheme,
+	organizationName string,
+) *RunnerPoolReconciler {
+	return &RunnerPoolReconciler{
+		Client:           client,
+		Log:              log,
+		Scheme:           scheme,
+		organizationName: organizationName,
+	}
+}
+
+//+kubebuilder:rbac:groups=actions.cybozu.com,resources=runnerpools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=actions.cybozu.com,resources=runnerpools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *RunnerPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("start reconciliation loop", "name", req.NamespacedName)
+
+	rp := &actionsv1alpha1.RunnerPool{}
+	if err := r.Get(ctx, req.NamespacedName, rp); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("runnerpool is not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "unable to get RunnerPool", "name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if rp.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(rp, runnerPoolFinalizer) {
+			err := r.addFinalizer(ctx, rp)
+			if err != nil {
+				r.Log.Error(err, "failed to add finalizer", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+			// Result does not change even if not requeued here.
+			// This just breaks down one reconciliation loop into small steps for simplicity.
+			r.Log.Info("added finalizer", "name", req.NamespacedName)
+			return ctrl.Result{Requeue: true}, nil
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(rp, runnerPoolFinalizer) {
+			err := r.cleanUpOwnedResources(ctx, req.NamespacedName)
+			if err != nil {
+				r.Log.Error(err, "failed to clean up deployment", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+
+			err = r.removeFinalizer(ctx, rp)
+			if err != nil {
+				r.Log.Error(err, "failed to remove finalizer", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("removed finalizer", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, d, func() error {
+		return r.updateDeploymentWithRunnerPool(rp, d)
```

Comment:
Hey! Just trying to learn something from similar projects as the maintainer of [another one](https://github.com/summerwind/actions-runner-controller) here-

With this architecture, how would you prevent a "busy" runner pod from being disrupted when you've triggered a rolling-update on the deployment?

For example, do we need something like our own finalized for pods(haven't tried it myself) that defers the pod deletion until the progressing workflow job completes?

Reply:
Hi, thanks for the other day on [this PR](https://github.com/summerwind/actions-runner-controller/pull/355).

This system does not block rolling updates of Pods even in the "busy" state for simplicity, because the
timing of the update can be controlled by ourselves, say, if we would like to update a Pod with a new container,
we just update it at night. Otherwise, we can upload a container with the same name but a different hash ID and
set `imagePullPolicy` as `Always`. `Always` is not a recommended setting for production, but this runs only on a
staging environment, so I believe that they accept such a configuration.

Reply:
Finalizer might help you implement such a behavior, but not sure if that is the best way to achieve it...

Reply:
Gotcha! Thanks for clarifying.

It would be okay that your "normal" update is being scheduled at night. However, I suppose you may still end up deploying in the evening due to hotfixes, which can result in fixing affected workflow runs while disrupting progressing workflow runs on the other hand.

It can be rooted in the definition of "minimal" or depend on your plan- but I'd still say that the ability to defer stopping busy runners would be "a great thing to have from day one". Just my two cents.

Reply:
Yes, this is a minimal implementation, so I'm not gonna do that in this task,
but I would do that once I ran this system for a while, and found that it was
absolutely necessary for us.

Anyway, thank you for getting touch with me at the very begining of this project!!

Reply:
OK! Best wishes to the future of this project and also- Thank you too for sharing this project and giving me inspiration ☺️